Enable receiving a rest config on TestSuite

alifelan opened the pull request:
Signed-off-by: Ali Felan <alifelan@google.com>

**What this PR does / why we need it**:
This adds an extra option to TestSuite, which allows KUTTL to run with a rest config, opening possibilities for customization on Go.

This is still missing testing and documentation, but wanted to share it to get feedback. My first try was making the config an exported name in harness, but that required a lot of changes in names (config to Config and Config() to GetConfig()). Adding it to TestSuite was easier, but it's the only non primitive type there.

Fixes #347

kensipe commented:
interesting @alifelan and thank you!   Apologies for missing the submitted issue which provided some of the context.  let's continue a conversation on the issue and return here.   It's a fairly small PR.. just need to understand the consequences of it.  Does this PR resolve the env issue for you?

alifelan commented:
Hey @kensipe! No problem, hope it makes more sense with the context in the issue, the PR is actually really small. But yeah, by doing this we can have KUTTL connect to the cluster using the rest config.

kensipe commented on pkg/test/harness.go:
```diff
@@ -236,6 +236,11 @@ func (h *Harness) Config() (*rest.Config, error) {
 		return h.config, nil
 	}
 
+	if h.TestSuite.Config != nil {
```
while this looks similar to the quick exit above...  it isn't the same IMO.   I would rather the `switch` below be the ways in which the `h.config` is configured... more importantly we should `h.T.Log` this.

kensipe reviewed (commented):
@alifelan ok... enough thinking... more doing (on my part).  I can't think of an issue here... it is an optional element.     As noted below,  I have a preference to have config selection fall into the same logic.. additionally this will not immediately return but will create a `kubeconfig` file which is used for some of the commands if included in the kuttl tests.

It would be great to have a test for this need feature as well.   
This change + a test for the new configuration and we can land this.

Is that something you are up for?  or should I merge and create a PR to cover it?

alifelan commented:
Sounds great! I updated it, now the check is in the switch, and also added a test. I'm looking for feedback on that, since I tried creating a test environment using `testutils.StartTestEnvironment`, and from there passing the config to harness. The problem with that is that the mocked controlplane is skipped on the service account check, otherwise raises an error. To achieve that, I set `StartControlPlane` to true and changed the order of the switch. I don't feel that's the perfect solution, since setting `StartControlPlane` in there is a bit of a lie, but I didn't want to add a config just for this specific use case. If you have any other solution, please let me know!

kensipe reviewed (approved):
/lgtm

Thanks so much!

kensipe commented:
the CI doesn't seem to run for non-committers... I don't remember the details around this... all standard CI activities tested good.  Ready to merge.   Adding a todo to enable CI for all PRs

kensipe commented:
For awareness.. this broken main on merge...  I should have caught it on review... but I blame the lack of CI builds for non-committers as the issue.   The `make generate` is called along with `verify-generate.sh` in the CI.  It rarely changes and is rarely called by developers.   We need the safety net of the CI.   I will submit a PR fix for the break.

kensipe commented:
@alifelan any thoughts on re-attempting this PR?   I hated reverting it.. but need main to be buildable.

alifelan commented:
Hey @kensipe! I've been taking a look, though I don't have a solution yet. Seems like `TLSClientConfig` is a promoted field in `rest.Config`, and it's `DeepCopy` method is the one being used (`rest.Config` has a `CopyConfig` method, but that won't be used by the deepcopy-gen). The only idea I have is moving the field from TestSuite to Harness, but I don't think that's the way to go from a design point of view. I'll think about it, and share my ideas with you!
